### PR TITLE
Fix/in-transit

### PR DESCRIPTION
**Pull request by kiwiidb**
Resolves https://github.com/getAlby/lndhub.go/issues/97

This PR will handle in transit payments, with the exceptions of some edge-cases listed below.

It will do this by always committing the outgoing transaction before trying to pay, and then possibly refunding the user in case of a payment failure. In case of payment failure, the outgoing invoice should also be updated with the failure reason.

The edge cases that should be handled in a later PR:
- LNDhub restarts while a payment is in transit. We can handle this by checking pending payments at startup and also handling any failed payment here.
- LND restarts while payment is pending, or there is a network failure while the payment is pending. This is a more difficult scenario to handle, and I am not sure how we should do this.

### CODE REVIEW

File: lib/service/invoices.go [hunk 1]
```diff
@@ -159,35 +158,29 @@ func (svc *LndhubService) PayInvoice(ctx context.Context, invoice *models.Invoic
...
-		paymentResponse, err = svc.SendInternalPayment(ctx, &tx, invoice)
+		paymentResponse, err = svc.SendInternalPayment(context.Background(), invoice)
 		if err != nil {
-			tx.Rollback()
+			svc.HandleFailedPayment(context.Background(), invoice, entry, err)
 			return nil, err
 		}
 	} else {
-		paymentResponse, err = svc.SendPaymentSync(ctx, &tx, invoice)
+		paymentResponse, err = svc.SendPaymentSync(context.Background(), invoice)
```
Comment (skosito) [inline lib/service/invoices.go:181]: here context.background was missing, also on handleSuccessfulPayment, added it on both places
does it make sense to use same context for couple of these or new one for all?

Comment (kiwiidb) [reply lib/service/invoices.go:181]: I think it doesn't really matter, we can use a new one to make it more clear. It's an empty context anyway.

File: lib/service/invoices.go [hunk 2]
```diff
@@ -196,23 +189,40 @@ func (svc *LndhubService) PayInvoice(ctx context.Context, invoice *models.Invoic
...
 	if err != nil {
-		tx.Rollback()
-		return nil, err
+		// TODO: error logging
+		return err
 	}
 
-	// Commit the DB transaction. Done, everything worked
-	err = tx.Commit()
+	// TODO: maybe save errors on the invoice?
```
[the remark this answers is not in the document]
Comment (kiwiidb) [reply lib/service/invoices.go:211]: Yeah, I think you can add a field for the error here, doesn't need to be a seperate PR.

[the remark this answers is not in the document]
Comment (kiwiidb) [reply lib/service/invoices.go:211]: The relevant issue is https://github.com/getAlby/lndhub.go/issues/106.

**Review by kiwiidb — COMMENTED** [review]
Nice work! Could you add another test for the following:

- A mock LND wrapper that has an internal channel. When SendPaymentSync is called, it waits for a message on this channel, if it receives the message it will fail the payment.
- Make an external payment, but give the request a context that is canceled after some time.
- Check the balance and assert the payment value is deducted
- Send a message to the channel so the payment fails
- Check the balance again and assert that nothing has been deducted.

I think if we have this test then the PR is ok to merge.

File: lib/service/invoices.go [hunk 3]
```diff
@@ -196,23 +189,46 @@ func (svc *LndhubService) PayInvoice(ctx context.Context, invoice *models.Invoic
...
+		CreditAccountID: entryToRevert.DebitAccountID,
+		DebitAccountID:  entryToRevert.CreditAccountID,
+		Amount:          invoice.Amount,
+	}
+	_, err := svc.DB.NewInsert().Model(&entry).Exec(ctx)
 	if err != nil {
-		tx.Rollback()
-		return nil, err
+		svc.Logger.Errorf("Could not insert transaction entry user_id:%v invoice_id:%v", invoice.UserID, invoice.ID)
+		return err
```
Comment (bumi) [inline lib/service/invoices.go:208]: those errors are still sent to Sentry? or should we do that explicitly here? 
We say above we run those in a Background context to ignore the request context. 

I just think we have to make sure that we have notifications for such exceptions. This one would require some admin to do something.

File: lib/service/invoices.go [hunk 4]
```diff
@@ -196,23 +189,46 @@ func (svc *LndhubService) PayInvoice(ctx context.Context, invoice *models.Invoic
...
-	// Commit the DB transaction. Done, everything worked
-	err = tx.Commit()
+	invoice.State = common.InvoiceStateError
+	if failedPaymentError != nil {
+		invoice.ErrorMessage = failedPaymentError.Error()
+	}
 
+	_, err = svc.DB.NewUpdate().Model(invoice).WherePK().Exec(ctx)
 	if err != nil {
-		return nil, err
```
Comment (bumi) [inline lib/service/invoices.go:212]: same here: sentry notification?

File: lib/service/invoices.go [hunk 5]
```diff
@@ -196,23 +189,46 @@ func (svc *LndhubService) PayInvoice(ctx context.Context, invoice *models.Invoic
...
+}
 
-	return &paymentResponse, err
+func (svc *LndhubService) HandleSuccessfulPayment(ctx context.Context, invoice *models.Invoice) error {
+	invoice.State = common.InvoiceStateSettled
+	invoice.SettledAt = schema.NullTime{Time: time.Now()}
+
+	_, err := svc.DB.NewUpdate().Model(invoice).WherePK().Exec(ctx)
+	if err != nil {
+		svc.Logger.Errorf("Could not update sucessful payment invoice user_id:%v invoice_id:%v", invoice.UserID, invoice.ID)
```
Comment (bumi) [inline lib/service/invoices.go:229]: sentry notification?

**Comment by kiwiidb** [comment]
Looks good to me, but @bumi must approve because I started the PR.

**Review by bumi — APPROVED** [review]
looks good!

**Pull request merged**